### ISSUE TITLE
Evaluate dev set in Translate.py: the decoder stills fetches from decoder_inputs, instead of previous predictions.

### ISSUE THREAD

**Issue by LittleYUYU** [tensorflow/tensorflow#1879]
In the training process defined in Translate.py, the model is created with `model = create_model(sess, False)` where `forward_only=False` means every step of decoding the decoder would fetch from decoder_inputs, instead of previous prediction. 

That's the same for evaluation of dev set, since the command 
`_, eval_loss, _ = model.step(sess, encoder_inputs, decoder_inputs, target_weights, bucket_id, True)` 
with `forward_only=True` actually doesn't change the decoding way. (It just decides the backward.)

So I guess the evaluation of dev set is not a real evaluation? Or is it a trick? I guess the correct way should be creating a model with current checkpoint params and set `forward_only=True`.

Thanks.

**Comment by lukaszkaiser** [comment]
It's exactly as you say! I think it's a bit too strong to call it "not real" -- the reported perplexities come from teacher-forced examples, and so are indeed not fully representative for what you'll get during decoding. But some people report this kind of perplexities in their papers and they are often a good measure to judge how the training process is progressing. In particular, I think they are enough to determine if the model is overfitting or not and whether it is converging or not. You could create a second model with the same parameters and `forward_only=True` and use that for decoding, but it'd use a lot of memory, that's why it's not done by default.

To get a final measure of the model quality, it's best to take a separate test set, decode it using the `decode` function [https://github.com/tensorflow/tensorflow/blob/master/tensorflow/models/rnn/translate/translate.py#L208] (which uses `forward_only=True` of course) and then use any measure on the result, e.g. BLEU. I'd also recommend to use a test set that's disjoint from the dev set for that, as you might be tuning hyper-parameters on the dev set.
